### PR TITLE
[quant] EqualizationQConfig to distinguish input/output activations

### PR DESCRIPTION
**Pull request by angelayi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59953 [quant] Equalization Observer modifications
* #59747 [quant] Input Weight Equalization - prepare modifications
* **#59739 [quant] EqualizationQConfig to distinguish input/output activations**

Summary: Created an EqualizationQConfig specifically for equalization.
This inherits from QConfig and is used to distinguish between inserting
an input observer with an output observer. Since the output observer
field is included in the EqualizationQConfig, we no longer need an
output observer field in the _InputEqualizationObserver

Test Plan: compiles

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29135298](https://our.internmc.facebook.com/intern/diff/D29135298)

### CODE REVIEW

**Comment by supriyar** [comment]
looks good. I think we may want to eventually move this into the qconfig.py file when it is more stable. For now seems like this is a good place as we develop this feature.

File: torch/quantization/fx/_equalize.py [hunk 1]
```diff
@@ -229,3 +225,41 @@ def calculate_equalization_scale(input_obs: _InputEqualizationObserver,
...
+
+    Observer classes have usually reasonable default arguments, but they can be
+    overwritten with `with_args` method (that behaves like functools.partial):
+
+    my_qconfig = EqualizationQConfig(input_activation=_InputEqualizationObserver.with_args(dtype=torch.qint8),
+                                    output_activation=MinMaxObserver.with_args(dtype=torch.qint8),
+                                    weight=_WeightEqualizationObserver.with_args(dtype=torch.qint8))
+    """
+    def __new__(cls, input_activation=torch.nn.Identity, output_activation=torch.nn.Identity, weight=torch.nn.Identity):
+        # catch common mistakes
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:250]: nit: remove this comment?

File: torch/quantization/fx/_equalize.py [hunk 2]
```diff
@@ -229,3 +225,41 @@ def calculate_equalization_scale(input_obs: _InputEqualizationObserver,
...
+        if isinstance(input_activation, nn.Module) or isinstance(output_activation, nn.Module) or isinstance(weight, nn.Module):
+            raise ValueError("EqualizationQConfig received observer instance, please pass observer class instead. " +
+                             "Use MyObserver.with_args(x=1) to override arguments to constructor if needed")
+        self = super(EqualizationQConfig, cls).__new__(cls, output_activation, weight)
+        self.input_activation = input_activation
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:255]: Should we also assert on the type of input_activation and weight? For equalization we should only support _InputEqualizationObserver and similar for weight

File: torch/quantization/fx/_equalize.py [hunk 3]
```diff
@@ -229,3 +225,41 @@ def calculate_equalization_scale(input_obs: _InputEqualizationObserver,
     equalization_scale = torch.sqrt((max_weights - min_weights) / (max_inputs - min_inputs))
 
     return equalization_scale
+
+
+class EqualizationQConfig(QConfig):
+    """
+    Describes how to quantize a layer or a part of the network specifically for
+    equalization by providing settings (observer classes) for inputs, outputs,
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:233]: Let's say input-weight equalization in the description to be more specific

File: torch/quantization/fx/_equalize.py (hunk 2 above)
Comment (angelayi) [reply torch/quantization/fx/_equalize.py:255]: I'm not sure how to check the type of the input_activation/weights as they are `functool.partial`s. I could do `isinstance(input_activation(), _InputEqualizationObserver)`, but that would create an `input_activation` instance.

**Review by vkuzo — APPROVED** [review]

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29135298).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29135298).

**Pull request closed without merging**